src/parse-nm.c

```diff
@@ -64,17 +65,306 @@ ap_type_from_str(const char* type_str)
+                g_debug("processing route_option: %s", split[i]);
+                gchar **kv = g_strsplit(split[i], "=", 2);
+                if (g_strcmp0(kv[0], "onlink") == 0)
+                    route->onlink = (g_strcmp0(kv[1], "true") == 0);
+                else if (g_strcmp0(kv[0], "initrwnd") == 0)
+                    route->advertised_receive_window = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "initcwnd") == 0)
+                    route->congestion_window = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "mtu") == 0)
+                    route->mtubytes = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "table") == 0)
+                    route->table = strtoul(kv[1], NULL, 10);
+                else if (g_strcmp0(kv[0], "src") == 0)
+                    route->from = g_strdup(kv[1]); //no need to free, will stay in netdef
+                else
+                    unhandled_data = TRUE;
+                g_strfreev(kv);
+            }
+            g_strfreev(split);
+
+            if (!unhandled_data)
+                _kf_clear_key(kf, group, options_key);
+            g_free(options_key);
+            g_free(options_kf_value);
+        }
+
+        /* Add route to array, clear keyfile */
+        g_array_append_val(*routes_arr, route);
+        if (!unhandled_data)
+            _kf_clear_key(kf, group, key);
+        g_free(key);
+        g_free(kf_value);
+    }
+}
+
+static void
+parse_dhcp_overrides(GKeyFile* kf, const gchar* group, NetplanDHCPOverrides* dataptr)
+{
+    if (   g_key_file_get_boolean(kf, group, "ignore-auto-routes", NULL)
+        && g_key_file_get_boolean(kf, group, "never-default", NULL)) {
+        (*dataptr).use_routes = FALSE;
+        _kf_clear_key(kf, group, "ignore-auto-routes");
+        _kf_clear_key(kf, group, "never-default");
+    }
+    handle_generic_uint(kf, group, "route-metric", &(*dataptr).metric, NETPLAN_METRIC_UNSPEC);
+}
+
+static void
+parse_search_domains(GKeyFile* kf, const gchar* group, GArray** domains_arr)
+{
+    g_assert(domains_arr);
+    gsize len = 0;
+    gchar **split = g_key_file_get_string_list(kf, group, "dns-search", &len, NULL);
+    if (split) {
+        if (len == 0) {
+            _kf_clear_key(kf, group, "dns-search");
+            return;
+        }
+        if (!*domains_arr)
+            *domains_arr = g_array_new(FALSE, FALSE, sizeof(char*));
+        for(unsigned i = 0; split[i]; ++i) {
+            char* s = g_strdup(split[i]); //no need to free, will stay in netdef
+            g_array_append_val(*domains_arr, s);
+        }
+        _kf_clear_key(kf, group, "dns-search");
+        g_strfreev(split);
+    }
+}
+
+static void
+parse_nameservers(GKeyFile* kf, const gchar* group, GArray** nameserver_arr)
+{
+    g_assert(nameserver_arr);
+    gchar **split = g_key_file_get_string_list(kf, group, "dns", NULL, NULL);
+    if (split) {
+        if (!*nameserver_arr)
```

Comment:
Related to the above: here we don't actually get or check the length of the resulting list - is this simply a different situation here and not being possible to be `len == 0`?

Reply:
Yeah it's a bit different, as NM generates a "dns-search=" line by default that we want to clear. It does not generate a "dns=" line by default. I restricted myself to only implicitly clearing the default values from the keyfile, but no (empty) data that was put in there manually (probably on purpose).